contracts/KSElasticLMV2.sol

```diff
@@ -0,0 +1,1015 @@
+    /// @param curSumRewardPerLiquidities current sumRewardPerLiquidities of farm, indexing by reward
+    function _updateRewardInfo(
+        uint256 nftId,
+        uint256 liquidity,
+        uint256[] memory curSumRewardPerLiquidities
+    ) internal {
+        uint256 length = curSumRewardPerLiquidities.length;
+        for (uint256 i; i < length; ) {
+            if (liquidity != 0) {
+                //calculate rewardAmount by formula rewardAmount = (sumRewardPerLiq - lastSumRewardPerLiq) * stake.liquidiy
+                uint256 rewardAmount = LMMath.calcRewardAmount(
+                    curSumRewardPerLiquidities[i],
+                    stakes[nftId].lastSumRewardPerLiquidity[i],
+                    liquidity
+                );
+
+                //accumulate reward into stake rewards
+                if (rewardAmount != 0) {
+                    stakes[nftId].rewardUnclaimed[i] += rewardAmount;
+                }
+            }
+
+            //store new sumReward into stake
+            stakes[nftId].lastSumRewardPerLiquidity[
+                    i
+                ] = curSumRewardPerLiquidities[i];
+
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    /// @dev update farm's sumRewardPerLiquidity
+    /// @dev update farm's lastUpdatedTime
+    /// @dev if block.timestamp > farm's endTime then update phase to settled
+    /// @param fId farm's id
+    /// @return curSumRewardPerLiquidity array of sumRewardPerLiquidity until now
+    function _updateFarmSumRewardPerLiquidity(
+        uint256 fId
+    ) internal returns (uint256[] memory curSumRewardPerLiquidity) {
+        uint256 length = farms[fId].phase.rewards.length;
+        curSumRewardPerLiquidity = new uint256[](length);
+
+        uint32 lastTouchedTime = farms[fId].lastTouchedTime;
+        uint32 endTime = farms[fId].phase.endTime;
+        bool isSettled = farms[fId].phase.isSettled;
+
+        for (uint256 i; i < length; ) {
+            uint256 preSumRewardPerLiquidity = farms[fId].sumRewardPerLiquidity[
+                i
+            ];
+
+            //calculate sumReward from lastTouchedTime until now
+            curSumRewardPerLiquidity[i] = _calcSumRewardPerLiquidity(
+                farms[fId].phase.rewards[i].rewardAmount,
+                farms[fId].phase.startTime,
+                endTime,
+                lastTouchedTime,
+                farms[fId].liquidity,
+                isSettled,
+                preSumRewardPerLiquidity
+            );
+
+            //if there is something changes, update into storage
+            if (curSumRewardPerLiquidity[i] != preSumRewardPerLiquidity)
+                farms[fId].sumRewardPerLiquidity[i] = curSumRewardPerLiquidity[
+                    i
+                ];
+
+            unchecked {
+                ++i;
+            }
+        }
+
+        //update farm lastTouchedTime, if passed endTime, update phase to settled
+        if (block.timestamp > lastTouchedTime)
```

Comment:
local variable here?

Reply:
yeah it's local variable, loaded from the beginning of this func

Reply:
I mean should we use local var for block.timestamp as it is called multiple times

Reply:
yeah I can cache it to local variable but it is needed? I think gas consume would be the same